pkg/client/event_tracking.go

```diff
@@ -0,0 +1,138 @@
+// process completes (no matter if it succeeded or failed), it should be removed
+// from this struct.
+type keyGenKeepTrack struct {
+	*keepEventTrack
+}
+
+// closeKeepTrack is used to track keep close events. It is used to ensure
+// that the process execution is not duplicated, e.g. when the client receives
+// the same event multiple times. When event is received, it should be noted
+// in this struct. When the signer generation process completes (no matter
+// if it succeeded or failed), it should be removed from this struct.
+type closeKeepTrack struct {
+	*keepEventTrack
+}
+
+// terminateKeepTrack is used to track keep terminate events. It is used to
+// ensure that the process execution is not duplicated, e.g. when the client
+// receives the same event multiple times. When event is received, it should be
+// noted in this struct. When the signer generation process completes (no matter
+// if it succeeded or failed), it should be removed from this struct.
+type terminateKeepTrack struct {
+	*keepEventTrack
+}
+
+// keepEventTrack is a simple event track implementation allowing to track
+// events that are happening once in the entire keep history, for example:
+// - keep opened, generating signer
+// - keep getting closed
+// - keep getting archived
+//
+// Those events do not require any special tracking or additional fields to
+// distinguish them. It is used to ensure that the process execution is not
+// duplicated, e.g. when the client receives the same event multiple times. When
+// event is received, it should be noted in this struct. When the signer
+// generation process completes (no matter if it succeeded or failed), it should
+// be removed from this struct.
+type keepEventTrack struct {
+	data  map[string]bool // <keep, bool>
+	mutex *sync.Mutex
+}
+
+func (ket *keepEventTrack) add(keepAddress common.Address) bool {
+	ket.mutex.Lock()
+	defer ket.mutex.Unlock()
+
+	if ket.data[keepAddress.String()] == true {
+		return false
+	}
+
+	ket.data[keepAddress.String()] = true
+
+	return true
+}
+
+func (ket *keepEventTrack) has(keepAddress common.Address) bool {
+	ket.mutex.Lock()
+	defer ket.mutex.Unlock()
+
+	return ket.data[keepAddress.String()]
+}
+
+func (ket *keepEventTrack) remove(keepAddress common.Address) {
+	ket.mutex.Lock()
+	defer ket.mutex.Unlock()
+
+	delete(ket.data, keepAddress.String())
+}
+
+// requestedSignaturesTrack is used to track signature calculation started after
+// signature request event is received. It is used to ensure that the process execution
+// is not duplicated, e.g. when the client receives the same event multiple times.
+// When event is received, it should be noted in this struct. When the signature
+// calculation process completes (no matter if it succeeded or failed), it should
+// be removed from this struct.
+type requestedSignaturesTrack struct {
+	data  map[string]map[string]bool // <keep, <digest, bool>>
+	mutex *sync.Mutex
```

Comment:
Why do we use a mutex pointer here?

Reply:
Explained [in the previous comment](https://github.com/keep-network/keep-ecdsa/pull/671/files/df9623df0bba07ebf4119a55cfaaefb3dc455f38#diff-eed7bc28eaadcd6dba03386d92e97ad8a4a42c524220bc645fc9135d4c37508c).